snakefile

```diff
@@ -0,0 +1,495 @@
+        old=rules.download_old_clades.output ,
+        new=rules.run_nextclade.output
+    output:
+        "data/{database}/nextclade.tsv"
+    shell:
+        "./bin/join-rows {input.new} {input.old} -o {output}"
+
+
+rule join_metadata_and_clades :
+    input:
+        clades = rules.join_clades.output ,
+        meta = "data/{database}/metadata.noClade.tsv"
+    output:
+        "data/{database}/metadata.tsv"
+    shell:
+        "./bin/join-metadata-and-clades {input.meta} {input.clades} -o {output}"
+
+
+###################################################################################
+## nCoV workflow helpers and files for slack
+###################################################################################
+
+rule flag_metadata :
+    input :
+        rules.join_metadata_and_clades.output
+    output :
+        "data/{database}/flagged_metadata.txt"
+    shell:
+        "./bin/flag-metadata {input} > {output}"
+
+rule check_locations :
+    input :
+        rules.join_metadata_and_clades.output
+    params:
+        idcolumn=lambda wildcards : config['idcolumn'][wildcards.database]
+    output :
+        "data/{database}/location_hierarchy.tsv"
+    shell:
+        "./bin/check-locations {input} {output} {params.idcolumn}"
+
+rule metadata_change:
+    input:
+        metadata = "data/{database}/metadata.tsv",
+    output:
+        "logs/{database}_metadata_change.msg"
+    params:
+        destination_metadata = lambda w: _get_S3_DST(w)+f"/{w.database}_metadata.tsv.gz",
+        idcolumn=lambda w : config['idcolumn'][w.database],
+        slack_channel = _get_slack_channel
+    shell:
+        """
+            # notify and upload metadata change
+
+            dst_local="$(mktemp -t metadata-XXXXXX.tsv)"
+            diff="$(mktemp -t metadata-changes-XXXXXX)"
+            additions="$(mktemp -t metadata-additions-XXXXXX)"
+            trap "rm -f '$dst_local' '$diff' '$additions'" EXIT
+
+            ./bin/compute-metadata-change {input.metadata} "{params.destination_metadata}" {params.idcolumn} $dst_local $diff $additions
+
+
+            # csv-diff outputs two newlines which -n ignores but -s does not
+            if [[ -n "$(< "$diff")" ]]; then
+                # "Notifying Slack about metadata change."
+                ./bin/notify-slack --upload "metadata-changes.txt" $SLACK_TOKEN {params.slack_channel} < "$diff"
+            else
+                echo "No metadata change."
+            fi
+            # checking additions
+            if [[ -s "$additions" ]]; then
+                # "Notifying Slack about metadata additions."
+                ./bin/notify-slack --upload "metadata-additions.tsv" $SLACK_TOKEN {params.slack_channel} < "$additions"
+
+                if [[ "{params.idcolumn}" == "gisaid_epi_isl" ]]; then
+                    ./bin/notify-users-on-new-locations "$additions" --slack-token $SLACK_TOKEN --slack-channel {params.slack_channel}
+                fi
+            fi
+            touch {output}
```

Comment:
It's not clear to me why it's beneficial to pull out the notification logic from the computation in this rule and the others like it. That is, why change `notify-on-metadata-change` to `compute-metadata-change` + the embedded shell in this rule? It's particularly odd to me because the computation is solely for the sake of the notification (right?), so this change ends up separating the two pieces of tightly-coupled code.

Reply:
The idea underlying this choice was to decouple creation of the message from sending the message in order to be more explicit. This being said, I see how this could be "beautified" by wrapping this in more bash scripts.